module/zfs/arc.c

```diff
@@ -3914,11 +3909,13 @@ arc_evict_hdr(arc_buf_hdr_t *hdr, uint64_t *real_evicted)
 	ASSERT(MUTEX_HELD(HDR_LOCK(hdr)));
 	ASSERT(HDR_HAS_L1HDR(hdr));
 	ASSERT(!HDR_IO_IN_PROGRESS(hdr));
+	ASSERT0(hdr->b_l1hdr.b_bufcnt);
+	ASSERT3P(hdr->b_l1hdr.b_buf, ==, NULL);
+	ASSERT0(zfs_refcount_count(&hdr->b_l1hdr.b_refcnt));
```

Comment:
What do you think about making some or all of these VERIFYs to make sure that the assumptions here are valid in non-debug builds?

Reply:
This is a very hot single-threaded piece of code, limiting system performance on some workloads.  I do not want assert impossible here without a good reason.  Any particular reason you think this may work different for debug and non-debug builds?

Reply:
If you're convinced that this can't happen then no need to be defensive. My concern was that if we find that the `b_buf` is not being destroyed because of some missed code path, then we end up with a memory leak which would be hard to track down. Doing a simple `VERIFY3P(hdr->b_l1hdr.b_buf, ==, NULL)` would allow for a quick diagnosis on non-debug bits. If you've stressed tested this significantly then I would agree that leaving these off makes sense if they can't happen.

Reply:
I've added the assertions to protect from some logical bugs.  Since it didn't fire yet on any tests, I tend to think it won't, and the assertions will just make sure that whoever modify the code in future will have a guard rail.

Reply:
And it likely won't leak any way.  It will get evicted to ghost state with the buffer, and on eviction from there arc_hdr_destroy() frees any possible remaining buffers.  I thought to remove it also, but appeared it is actually used when arc_buf_destroy() calls remove_reference() for anonymous header without getting the hash lock.

Reply:
Ok. The arc behaves very differently in non-debug builds so some of the inherit race conditions may not show up there. I will trust that you're confident enough in the change.